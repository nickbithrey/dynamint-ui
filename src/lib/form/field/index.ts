import { Field, IField, IFieldType, IFieldValueType, ITypedField } from './Field';

export {
    Field,
    IField,
    IFieldType, 
    IFieldValueType, 
    ITypedField
};