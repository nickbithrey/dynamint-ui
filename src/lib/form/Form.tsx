import * as React from 'react';
import { IButton, ButtonContainer } from 'lib/button';
import { Key } from '~/lib';
import StorageBuilder, { Storage } from 'service/storage';

export interface IForm {
    storageKey: string;
    entity: any;
    save: (obj: any) => void;
    back: () => void;
}

export interface IFormField {
    updateField: (name: string, value: any) => void;
    entity: any;
}

interface IFormState {
    [key: string]: any;
}

export class Form extends React.Component<IForm, IFormState> {
    
    storage: Storage<string, IFormState>;
    
    constructor(props: IForm) {
        super(props);
        this.state = {...props.entity};
    }
    
    componentDidMount() {
        this.storage = new StorageBuilder<string, IFormState>()
            .withKey('dynamint-ui-form-' + this.props.storageKey)
            .withAdaptor('localStorage')
            .build();
    }

    componentDidUpdate() {
        this.storage.store(this.state);
    }
    
    componentWillUnmount() {
        this.storage.clear();
    }
    
    updateField = (name: string, value: any): void => {
        this.setState({
            [name]: value
        });
    }
    
    save = (): void => {
        this.props.save(this.state);
        this.props.back();
    }
    
    buttons = (): Array<IButton & Key> => {
        return [
            {key: 'back', content: 'Back', type: 'default', onClick: this.props.back},
            {key: 'save', content: 'Save', type: 'primary', onClick: this.save}
        ];
    }
    
    render() {
        const {
            children,
            entity
        } = this.props;
        return (
            <div>
                { React.Children.map(children, child => React.cloneElement(child as React.ReactElement<any & IFormField>, { entity: entity, update: this.updateField }) ) }
                <ButtonContainer buttons={this.buttons()} />
            </div>
        );
    }
}