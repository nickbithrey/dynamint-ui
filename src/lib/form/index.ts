import { Form, IForm } from './Form';

export {
    Form,
    IForm
};