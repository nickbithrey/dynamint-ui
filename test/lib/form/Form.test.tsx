import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { Form, IForm } from 'lib/form/Form';

const FormContent = () => <p>form</p>;

describe('<Form />', () => {
   
    const props = {
        storageKey: 'key',
        entity: {
            reference: 'ref',
            desc: 'desc'
        },
        save: jest.fn(),
        back: jest.fn()
    };
    
    let wrapper: ShallowWrapper<Form>;
    
    beforeEach(() => {
        wrapper = shallow(<Form {...props}><FormContent /></Form>);
    });
    
    it('renders valid tag and matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    it('passes updateField prop to children', () => {
        expect(wrapper.find(FormContent).prop('update')).toBe((wrapper.instance() as Form).updateField);
    });
    
    it('will clear the window storage when unmounting', () => {
        const key = 'dynamint-ui-form-' + props.storageKey;
        window.localStorage.setItem(key, 'store');
        expect(window.localStorage.getItem(key)).toEqual('store');
        
        wrapper.unmount();
        
        expect(window.localStorage.getItem(key)).toBeNull();
    });
    
    it('will store the state to storage on update', () => {
        const key = 'dynamint-ui-form-' + props.storageKey;
        window.localStorage.removeItem(key);
        expect(window.localStorage.getItem(key)).toBeNull();
        wrapper.instance().componentDidUpdate({}, {});
        expect(window.localStorage.getItem(key)).not.toBeNull();
    });
    
    it('will initialise the storage on mount', () => {
        expect((wrapper.instance() as Form).storage).not.toBeNull();
    });
    
    it('sets the state on the update method', () => {
        expect(wrapper.state()).toEqual({
            reference: 'ref',
            desc: 'desc'
        });
        (wrapper.instance() as Form).updateField('name', 'value');
        expect(wrapper.state()).toEqual({
            reference: 'ref',
            desc: 'desc',
            name: 'value'
        });
    });
    
    it('calls save prop on save method', () => {
        const save: jest.MockedFunction<any> = props.save;
        expect(save.mock.calls).toHaveLength(0);
        (wrapper.instance() as Form).save();
        expect(save.mock.calls).toHaveLength(1);
        expect(save.mock.calls[0][0]).toEqual({
            reference: 'ref',
            desc: 'desc'
        });
    });
    
});
