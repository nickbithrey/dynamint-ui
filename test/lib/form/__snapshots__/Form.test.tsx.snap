// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> renders valid tag and matches snapshot 1`] = `
<div>
  <FormContent
    entity={
      Object {
        "desc": "desc",
        "reference": "ref",
      }
    }
    key=".0"
    update={[Function]}
  />
  <Component
    buttons={
      Array [
        Object {
          "content": "Back",
          "key": "back",
          "onClick": [MockFunction],
          "type": "default",
        },
        Object {
          "content": "Save",
          "key": "save",
          "onClick": [Function],
          "type": "primary",
        },
      ]
    }
  />
</div>
`;
